### PR TITLE
Draft: Improve radix sort's peer prefix algorithms

### PR DESCRIPTION
**Pull request by adamfidel**
This is the second attempt to merge these changes. See #680.

This PR reduces the peer prefix algorithms' iteration space from all radix states to just the number of radix bits for the subgroup_ballot and atomic_fetch_or helpers. This decreases the number of subgroup collectives during the peer prefix portion from 16 per input element to 4 per input element.

It also changes the private work-item-local array for offsets to a shared local array and rewrites the serial scan to be a group scan.

The first change (reduction in iteration space) relies on the second change (shared offsets), so this PR cannot be split into two separate PRs.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 1]
```diff
@@ -514,21 +579,22 @@ __radix_sort_reorder_submit(_ExecutionPolicy&& __exec, ::std::size_t __segments,
...
                 _PeerHelper __peer_prefix_hlp(__self_item, __peer_temp);
 
-                // 1. create a private array for storing offset values
-                //    and add total offset and offset for compute unit for a certain radix state
-                _OffsetT __offset_arr[__radix_states];
                 const ::std::size_t __scan_size = __segments + 1;
-                _OffsetT __scanned_bin = 0;
-                __offset_arr[0] = __offset_rng[__wgroup_idx];
-                for (::std::uint32_t __radix_state_idx = 1; __radix_state_idx < __radix_states; ++__radix_state_idx)
+                for (::std::uint32_t __radix_state = __self_lidx; __radix_state < __radix_states; __radix_state += __sg_size)
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:583]: 1) The changes are connected with the changes in __peer_prefix_helper class?
if not, may be it makes sense to do it in a separate PR?
2) And the second question  - how the changes (lines 583 - 596) influent to performance?

Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:583]: BTW, we allow that `__sg_size`  is sub group size here, right?
But if we look on the sycl::nd_range for re-order we may notice that we have the iteration  space `(__segments * __sg_size, __sg_size)`, where `__sg_size` is a work group size and `__segments` is number of work groups.
`__sg_size` is very small  (8 or 4 for int)... So, it looks like we have low GPU occupancy here... Or I am wrong?

Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:583]: For the first question, this PR is moving the loop over the radix states inside of the peer prefix class. This is because two of the peer prefix helpers (`atomic_ballot` and `fetch_or`) are now iterating over the radix bits instead of all of the radix states. I do not believe it makes sense to make a PR with just those changes since they are connected to the optimizations presented in this PR.

For the second question, the changes to store the offsets in local memory and perform a group scan in isolation creates a modest slowdown. However, storing the offsets in local memory and sharing them among the work-items is necessary for the changes to the `atomic_ballot` and `fetch_or` peer prefix helpers. Altogether, there is a performance speedup.

Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:583]: This loop is filling the offset array (which is size radix states) cooperatively amongst the work-items. For the case of 4 bits, we need to read 16 elements from the input offset range. If the subgroup size is less than 16 (such as 8 or 4 in your example), then each work-item is actually performing multiple trips through this loop and the occupancy is high (in the sense that each work-item is busy). It is only in the case where the subgroup size is larger than 16 that some work-items are left idle for this loop.

**Comment by adamfidel** [comment]
These changes no longer provide a sufficient enough performance improvement to justify merging.
